File: Graph/num-islands.js
/*

Given a 2d grid map of 1s (land) and 0s (water), count the number of islands.
An island is surrounded by water and is formed by connecting adjacent lands
horizontally or vertically. You may assume all four edges of the grid are all
surrounded by water.

EXAMPLE 1:

Input:
[
  [0, 1, 1, 1, 0],
  [1, 1, 0, 1, 0],
  [1, 1, 0, 0, 0],
  [0, 0, 0, 0, 0]
]
Output: 1

Input:
[
  [1, 1, 0, 0, 0],
  [1, 1, 0, 0, 0],
  [0, 0, 1, 0, 0],
  [0, 0, 0, 1, 1]
]
Output: 3

Assume that the grid will be an array of arrays of numbers either 0 or 1, and
that the grid will have at least one element.

//stepping stone into different graph questions. 



*/

const numIslands = grid => {
  //edge case: if grid is not defined, return 0
  if (!grid) return 0;
  //create a num islands counter
  let islands = 0;

  //create a grid to store visited information (T -> visited, F -> not visited)
  const seen = grid.map(row => row.map(col => false));
  
  for (let row = 0; row < grid.length; row++){
    for (let col = 0; col < grid[0].length; col++){
        if (grid[row][col] === 1 && !seen[row][col]){
          dfs(grid, seen, row, col);
          islands += 1;
        }
    }
  }
  return islands;
};

const dfs = (grid, seen, row, col) => {

  //base case: if row or column reaches lowest/highest index, if the position is already seen, or if the current element is a 0 -> return
  if (row < 0 || row >= grid.length || col < 0 || col >= grid[0].length || seen[row][col] || grid[row][col] === 0) return;
  //change the current element to be seen 
  seen[row][col] = true;

  dfs(grid, seen, row + 1, col);
  dfs(grid, seen, row - 1, col);
  dfs(grid, seen, row, col + 1);
  dfs(grid, seen, row, col - 1);
}





console.log(numIslands([
  [1, 1, 0, 0, 0],
  [1, 1, 0, 0, 0],
  [0, 0, 1, 0, 0],
  [0, 0, 0, 1, 1]
]))

module.exports = {numIslands};

File: Graph/num_islands.py
from typing import List


# class Solution:
#     def numIslands(self, grid: List[List[str]]) -> int:
#         count = 0
#         visited = [[False] * len(grid[0]) for _ in range(len(grid))]
        
#         for x in range(len(grid)):
#             for y in range(len(grid[0])):
                
#                 if grid[x][y] == "1" and not visited[x][y]:
#                     self.dfs(grid, visited, x, y)
#                     count += 1
                          
#         return count
    

#     def dfs(self, grid, visited, x, y):
        
#         if x < 0 or y < 0 or x >= len(grid) or y >= len(grid[0]) or grid[x][y] == "0" or visited[x][y]: 
#             return 0
#         if grid[x][y] == "1":
#             visited[x][y] = True

#             self.dfs(grid, visited, x + 1, y)  
#             self.dfs(grid, visited, x - 1, y) 
#             self.dfs(grid, visited, x, y + 1) 
#             self.dfs(grid, visited, x, y - 1)


         
        
        
        
grid = [
  ["1","1","1","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","0","0"]
]
solution = Solution()
print(solution.numIslands(grid))